drivers: ad559x: implement double range option

jeppenodgaard opened the pull request:
Add support for double range (2 x Vref) for both ADC and DAC.

Commit:
dts: bindings: adc: ad559x: add double range option

Add boolean option to use 2 x voltage reference as upper ADC input range.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

Commit:
drivers: adc: adc_ad559x: implement double range option

Optionally write to ad559x register bit which increases ADC input range
from 0V to 2 x Vref.

Create an ADC API per instance since the double range option and thus
`ref_internal` is instance specific.

The functionality has been tested via double range DAC and wiring IO1 and
IO3:
```
$ dac setup ad5593dac 1 12
$ adc ad5593adc resolution 12
$ adc ad5593adc channel id 3
$ dac write_value ad5593dac 1 0
$ adc ad5593adc read 3
read: 1
$ dac write_value ad5593dac 1 1024
$ adc ad5593adc read 3
read: 1021
$ dac write_value ad5593dac 1 2048
$ adc ad5593adc read 3
read: 2044
$ dac write_value ad5593dac 1 3072
$ adc ad5593adc read 3
read: 3069
$ dac write_value ad5593dac 1 4095
$ adc ad5593adc read 3
read: 4091
```

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

Commit:
dts: bindings: dac: ad559x: add double range option

Add boolean option to use 2 x voltage reference as upper DAC output range.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

Commit:
drivers: dac: ad559x: implement double range option

Optionally write to ad559x register bit which increases DAC output range
from 0V to 2 x Vref.

The functionality has been tested on AD5593r.
The voltage at 4095 has been measured to 5.01 volts.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

Commit:
drivers: ad559x: fix mfd_ad559x_i2c_read_reg

Reading a register requires 0x70 to be ORed with the register address.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

jeppenodgaard commented:
Rebase to restart CI which seemed stuck.

aasinclair commented on drivers/dac/dac_ad559x.c:
```diff
@@ -99,6 +113,7 @@ static int dac_ad559x_init(const struct device *dev)
 #define DAC_AD559X_DEFINE(inst)                                                                    \
 	static const struct dac_ad559x_config dac_ad559x_config##inst = {                          \
 		.mfd_dev = DEVICE_DT_GET(DT_INST_PARENT(inst)),                                    \
+		.range_double = DT_INST_PROP_OR(inst, range_double, false),                        \
```
Boolean properties don't need to use _OR.

```suggestion
		.range_double = DT_INST_PROP(inst, range_double),                        \
```

aasinclair commented on drivers/adc/adc_ad559x.c:
```diff
@@ -272,24 +286,29 @@ static int adc_ad559x_init(const struct device *dev)
+		.channel_setup = adc_ad559x_channel_setup,                                         \
+		.read = adc_ad559x_read,                                                           \
+		IF_ENABLED(CONFIG_ADC_ASYNC, (.read_async = adc_ad559x_read_async,))               \
+		.ref_internal = COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, range_double), (2), (1)) * \
+				AD559X_ADC_VREF_MV,                                                \
```
You can't use HAS_PROP with a boolean, maybe try something like:

```suggestion
		.ref_internal = AD559X_ADC_VREF_MV * (1 + DT_INST_PROP(inst, range_double)),       \

```

bbilas replied:
or 

<pre>
#if DT_INST_PROP(inst, range_double)
#define AD559X_REF_INTERNAL
	.ref_internal = AD559X_ADC_VREF_MV * 2,
#else
#define AD559X_REF_INTERNAL
	.ref_internal = AD559X_ADC_VREF_MV,
#endif
</pre>

aasinclair commented on drivers/adc/adc_ad559x.c:
```diff
@@ -272,24 +286,29 @@ static int adc_ad559x_init(const struct device *dev)
-#endif
-	.ref_internal = AD559X_ADC_VREF_MV,
-};
+/* clang-format off */
```
Why the clang-format disable?
If it is because of the weird formatting after IF_ENABLED, you could try moving this section to the end of the struct.

jeppenodgaard commented:
Thanks for the review @aasinclair.
I should have addressed your comments.
The clang-format is still not perfect, but now I think your suggestion and the format is better than disabling.

bbilas commented on include/zephyr/drivers/mfd/ad559x.h:
```diff
@@ -21,6 +21,8 @@ extern "C" {
 #define AD559X_REG_GPIO_SET       0x09U
 #define AD559X_REG_GPIO_INPUT_EN  0x0AU
 #define AD559X_REG_PD_REF_CTRL    0x0BU
+#define AD559X_REG_GPIO_OD_CONFIG 0x0CU
```
This is not related to this commit at all.

bbilas commented on drivers/dac/dac_ad559x.c:
```diff
@@ -88,6 +89,19 @@ static int dac_ad559x_init(const struct device *dev)
+		if (ret < 0) {
+			return ret;
+		}
+		ret = mfd_ad559x_write_reg(config->mfd_dev, AD559X_REG_GEN_CTRL, tmp);
```
How is it supposed to work? You do a read and write the same register without any modification.

bbilas commented on drivers/dac/dac_ad559x.c:
```diff
@@ -88,6 +89,19 @@ static int dac_ad559x_init(const struct device *dev)
+		ret = mfd_ad559x_read_reg(config->mfd_dev, AD559X_REG_GEN_CTRL, 0, &tmp);
+		if (ret < 0) {
+			return ret;
+		}
```
Add a `nl`.

jeppenodgaard commented:
Thanks for the review @bbilas and especially catching the missing register update.
I also changed it so the bit is always modified based on the dts property.
Otherwise the bit would not be cleared when `range-double` is not present in the dts.

bbilas reviewed (commented):
You have a typo in commit message

`s/referance/reference`

jeppenodgaard commented:
@bbilas I addressed the typo and your ideas, expect for the code that I already changed via @aasinclair's comment.

bbilas commented:
> @bbilas I addressed the typo and your ideas, expect for the code that I already changed via @aasinclair's comment.

That's okay. Please rename the `range-double` property with `double-input-range` (which IMO. is more understandable) as the last thing and we can approve this MR.

jeppenodgaard commented:
> That's okay. Please rename the `range-double` property with `double-input-range` (which IMO. is more understandable) as the last thing and we can approve this MR.

Changed to `double-input-range` for ADC and `double-output-range` for DAC.

bbilas commented on drivers/adc/adc_ad559x.c:
```diff
@@ -272,24 +290,31 @@ static int adc_ad559x_init(const struct device *dev)
+#else
+#define ADC_AD559X_ASYNC()
 #endif
-	.ref_internal = AD559X_ADC_VREF_MV,
```
I took a look at this part again and I'm worried why you didn't do it this way:

<pre>
static const struct adc_driver_api adc_ad559x_api = {
	.channel_setup = adc_ad559x_channel_setup,
	.read = adc_ad559x_read,
#ifdef CONFIG_ADC_ASYNC
	.read_async = adc_ad559x_read_async,
#endif
	.ref_internal = AD559X_ADC_VREF_MV * (1 + DT_INST_PROP(inst, double_input_range)),
};
</pre>

Then you could remove entire `ADC_AD559X_DRIVER_API` and etc...

jeppenodgaard replied:
Because different instances can have different `.ref_internal`.

bbilas replied:
You are right, I have forgotten about that.

ttmut commented on drivers/mfd/mfd_ad559x_i2c.c:
```diff
@@ -35,7 +37,9 @@ static int mfd_ad559x_i2c_read_reg(const struct device *dev, uint8_t reg, uint8_
 	ARG_UNUSED(reg_data);
 
-	__ASSERT((reg & 0xf0) == 0, "reg bits [7:4] should be 0: 0x%x", reg);
+	if (reg >= AD559X_REG_SEQ_ADC || reg <= AD559X_REG_IO_TS_CONFIG) {
+		reg |= AD559X_REG_RD_POINTER;
+	}
```
Just being curious, could you explain what this block does? I could not figure it out from the datasheet.

ttmut replied:
I was actually looking at AD5592R and that was what confused me. Thanks for the explanation.

jeppenodgaard commented:
> Thanks for the review @aasinclair. I should have addressed your comments. The clang-format is still not perfect, but now I think your suggestion and the format is better than disabling.

Ping @aasinclair
